ludwig/datasets/loaders/av_mnist.py

```diff
@@ -0,0 +1,214 @@
+            self.modal_separate = modal_separate
+            self.modal = modal
+        except ImportError:
+            logger.error(
+                "torchvision is not installed. "
+                "In order to install all image feature dependencies run "
+                "pip install ludwig[image]"
+            )
+            raise
+        super().__init__(config, cache_dir)
+
+    def transform_files(self, file_paths: List[str]) -> List[str]:
+        """
+        Args:
+            file_paths (List[string]): The directory where all the .npy files are located for the training and test data.
+        Returns:
+            The array of directory paths
+        """
+        self.process_source_dataset(self.raw_dataset_dir)
+        for dataset in ["training", "testing"]:
+            if not self.modal_separate:
+                if dataset == "training":
+                    self.audio_data = np.load(os.path.join(self.raw_dataset_dir, "audio", "train_data.npy"))
+                    self.mnist_data = np.load(os.path.join(self.raw_dataset_dir, "image", "train_data.npy"))
+                    self.labels = np.load(os.path.join(self.raw_dataset_dir, "train_labels.npy"))
+                else:
+                    self.audio_data = np.load(os.path.join(self.raw_dataset_dir, "audio", "test_data.npy"))
+                    self.mnist_data = np.load(os.path.join(self.raw_dataset_dir, "image", "test_data.npy"))
+                    self.labels = np.load(os.path.join(self.raw_dataset_dir, "test_labels.npy"))
+
+                self.audio_data = self.audio_data[:, np.newaxis, :, :]
+                self.mnist_data = self.mnist_data.reshape(self.mnist_data.shape[0], 1, 28, 28)
+                self.write_output_dataset(self.labels, self.audio_data, os.path.join(self.raw_dataset_dir, dataset))
+                self.write_output_dataset(self.labels, self.mnist_data, os.path.join(self.raw_dataset_dir, dataset))
+            else:
+                if self.modal:
+                    if self.modal not in ["audio", "image"]:
+                        raise ValueError("the value of modal is allowed")
+                    if dataset == "train":
+                        self.data = np.load(os.path.join(self.raw_dataset_dir, self.modal, "train_data.npy"))
+                        self.labels = np.load(os.path.join(self.raw_dataset_dir, "train_labels.npy"))
+                    else:
+                        self.data = np.load(os.path.join(self.raw_dataset_dir, self.modal, "test_data.npy"))
+                        self.labels = np.load(os.path.join(self.raw_dataset_dir, "test_labels.npy"))
+
+                    if self.modal == "audio":
+                        self.data = self.data[:, np.newaxis, :, :]
+                    elif self.modal == "image":
+                        self.data = self.data.reshape(self.data.shape[0], 1, 28, 28)
+                    self.write_output_dataset(self.labels, self.data, os.path.join(self.raw_dataset_dir, dataset))
```

Comment:
To avoid code duplication, would it be worth creating two private functions `_transform_and_load_image_files` and `_transform_and_load_audio_files` that look something like this:

```
def _transform_and_load_image_files(self, dataset: str) -> None:
    if dataset == "training":
        self.image_data = np.load(os.path.join(self.raw_dataset_dir, "image", "train_data.npy"))
    ....
       self.mnist_data = self.mnist_data.reshape(self.mnist_data.shape[0], 1, 28, 28)
       self.write_output_dataset(self.labels, self.mnist_data, os.path.join(self.raw_dataset_dir, dataset))
```

That way, we can just call these functions and greatly simplify `transform_files`

Reply:
Added